# Data Import & Data Cleaning

**Imports**

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

#Will need seaborn for visualizations and will need SciKit learn for modeling & metrics

**Read-in the training data**

In [3]:
train = pd.read_csv('./datasets/train.csv')

In [4]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2051 entries, 0 to 2050
Data columns (total 81 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Id               2051 non-null   int64  
 1   PID              2051 non-null   int64  
 2   MS SubClass      2051 non-null   int64  
 3   MS Zoning        2051 non-null   object 
 4   Lot Frontage     1721 non-null   float64
 5   Lot Area         2051 non-null   int64  
 6   Street           2051 non-null   object 
 7   Alley            140 non-null    object 
 8   Lot Shape        2051 non-null   object 
 9   Land Contour     2051 non-null   object 
 10  Utilities        2051 non-null   object 
 11  Lot Config       2051 non-null   object 
 12  Land Slope       2051 non-null   object 
 13  Neighborhood     2051 non-null   object 
 14  Condition 1      2051 non-null   object 
 15  Condition 2      2051 non-null   object 
 16  Bldg Type        2051 non-null   object 
 17  House Style   

### Initial observations:
- Total of 2,051 observations. 
- There are 80 total variables. 
- Most variables don't have null values
- Pool QC, Misc Feature, Fence, and Alley have significantly fewer variables
- Dataset contains both ID and PID, ID is the variable we need in the final dataset for identification

### Data Cleaning Goals:
1. Evaluate Variable Types: does the type make sense for each variable?
2. Examine variable value counts for categorical and ordinal variables
3. Drop Variables if they do not have values for enough observations
4. Examine variable 5 number summary for outlier values or typos for int & float variables

In [7]:
train['MS SubClass'].value_counts()
# This variable matches the documentation for the dataset
#This variable should be an object/ string this is a categorical variable

20     770
60     394
50     198
120    132
30     101
70      90
160     88
80      86
90      75
190     46
85      28
75      16
180     11
45      11
40       4
150      1
Name: MS SubClass, dtype: int64

In [8]:
train['MS Zoning'].value_counts()
# The variable counts do not match the documentation for this dataset
# I (all) is not an expected value
# As expected, there are no RP or residential park

RL         1598
RM          316
FV          101
C (all)      19
RH           14
A (agr)       2
I (all)       1
Name: MS Zoning, dtype: int64

In [10]:
train[train['MS Zoning']== "I (all)"]
#Don't see anything else about this observation that could indicate what I stands for...

Id        PID  MS SubClass MS Zoning  Lot Frontage  Lot Area Street  \
581  946  912251110           30   I (all)         109.0     21780   Grvl   

    Alley Lot Shape Land Contour  ... Screen Porch Pool Area Pool QC Fence  \
581   NaN       Reg          Lvl  ...            0         0     NaN   NaN   

    Misc Feature Misc Val Mo Sold Yr Sold  Sale Type  SalePrice  
581          NaN        0       3    2009      ConLD      57625  

[1 rows x 81 columns]

In [ ]:
train['MS Zoning'].value_counts()

In [11]:
train['Lot Frontage'].describe()
#missing values but variable checks out

count    1721.000000
mean       69.055200
std        23.260653
min        21.000000
25%        58.000000
50%        68.000000
75%        80.000000
max       313.000000
Name: Lot Frontage, dtype: float64

In [13]:
train['Lot Area'].describe()
#No missing values, large range

count      2051.000000
mean      10065.208191
std        6742.488909
min        1300.000000
25%        7500.000000
50%        9430.000000
75%       11513.500000
max      159000.000000
Name: Lot Area, dtype: float64

In [14]:
train['Street'].value_counts()
#Only 7 values are gravel... will this be a significant variable for analysis?

Pave    2044
Grvl       7
Name: Street, dtype: int64

In [15]:
train['Alley'].value_counts()
#Mostly missing, only applicable for houses with alleys

Grvl    85
Pave    55
Name: Alley, dtype: int64

In [16]:
train['Lot Shape'].value_counts()
#What does lot shape mean? 
#This variable and value counts match the documentation

Reg    1295
IR1     692
IR2      55
IR3       9
Name: Lot Shape, dtype: int64

In [17]:
train['Land Contour'].value_counts()
#no missing values
#values correspond to documentation

Lvl    1843
HLS      85
Bnk      80
Low      43
Name: Land Contour, dtype: int64

In [18]:
train['Utilities'].value_counts()
#No missing values
#Only 2 observations that are not allpub

AllPub    2049
NoSeWa       1
NoSewr       1
Name: Utilities, dtype: int64

In [20]:
train['Lot Config'].value_counts()
#No missing values
#Values match documentation

Inside     1503
Corner      348
CulDSac     131
FR2          60
FR3           9
Name: Lot Config, dtype: int64

In [21]:
train['Land Slope'].value_counts()
#No missing values & matches documentation
#Only a little variation in values

Gtl    1953
Mod      88
Sev      10
Name: Land Slope, dtype: int64

In [22]:
train['Neighborhood'].value_counts()

NAmes      310
CollgCr    180
OldTown    163
Edwards    143
Somerst    130
NridgHt    122
Gilbert    116
Sawyer     111
NWAmes      87
SawyerW     87
Mitchel     82
BrkSide     76
Crawfor     71
IDOTRR      69
NoRidge     48
Timber      48
StoneBr     38
SWISU       32
ClearCr     27
MeadowV     24
Blmngtn     22
BrDale      19
Veenker     17
NPkVill     17
Blueste      6
Greens       3
GrnHill      2
Landmrk      1
Name: Neighborhood, dtype: int64

In [23]:
train['Condition 1'].value_counts()

Norm      1767
Feedr      109
Artery      70
RRAn        36
PosN        27
RRAe        21
PosA        12
RRNn         6
RRNe         3
Name: Condition 1, dtype: int64

In [25]:
train['Condition 2'].value_counts()
#So few values non-Norm
#No missing values

Norm      2025
Feedr       11
Artery       5
PosA         3
PosN         3
RRNn         2
RRAe         1
RRAn         1
Name: Condition 2, dtype: int64

In [27]:
train['Bldg Type'].value_counts()
#No missing values and values track with documentation

1Fam      1700
TwnhsE     161
Duplex      75
Twnhs       69
2fmCon      46
Name: Bldg Type, dtype: int64

In [28]:
train['House Style'].value_counts()

1Story    1059
2Story     598
1.5Fin     218
SLvl        94
SFoyer      50
2.5Unf      14
1.5Unf      12
2.5Fin       6
Name: House Style, dtype: int64

In [29]:
train['Overall Qual'].describe()
#data ranges are as expected
#this variable should be an object not float unless i want to treat my ordinal variables that way

count    2051.000000
mean        6.112140
std         1.426271
min         1.000000
25%         5.000000
50%         6.000000
75%         7.000000
max        10.000000
Name: Overall Qual, dtype: float64

In [30]:
train['Overall Cond'].describe()
#What type of variable do i need this to be?

count    2051.000000
mean        5.562165
std         1.104497
min         1.000000
25%         5.000000
50%         5.000000
75%         6.000000
max         9.000000
Name: Overall Cond, dtype: float64

In [31]:
train['Year Built'].describe()
#looks fine- no weird outliers
#Should year be treated as a categorical/ ordinal variable or as a float/ int?

count    2051.000000
mean     1971.708922
std        30.177889
min      1872.000000
25%      1953.500000
50%      1974.000000
75%      2001.000000
max      2010.000000
Name: Year Built, dtype: float64

In [33]:
train['Year Remod/Add'].value_counts()
#This variable equals year built if there was no remodel or addition
#This variable looks fine

1950    262
2006    139
2007    123
2005    100
2004     83
       ... 
1986      9
2010      8
1983      7
1981      6
1982      3
Name: Year Remod/Add, Length: 61, dtype: int64

In [34]:
train['Roof Style'].value_counts()
#No missing values, values match documentation

Gable      1619
Hip         397
Flat         13
Gambrel      12
Mansard       7
Shed          3
Name: Roof Style, dtype: int64

In [35]:
train['Roof Matl'].value_counts()
#No missing values
#Not a lot of variation in values

CompShg    2025
Tar&Grv      15
WdShngl       5
WdShake       4
Membran       1
ClyTile       1
Name: Roof Matl, dtype: int64

In [36]:
train['Exterior 1st'].value_counts()
# No missing values
# Values match documentation

VinylSd    724
MetalSd    331
HdBoard    300
Wd Sdng    276
Plywood    152
CemntBd     90
BrkFace     64
WdShing     45
AsbShng     33
Stucco      27
BrkComm      3
Stone        2
CBlock       2
ImStucc      1
AsphShn      1
Name: Exterior 1st, dtype: int64

In [38]:
train['Exterior 2nd'].value_counts()
#Good variety
#No missing variables

VinylSd    721
MetalSd    324
HdBoard    275
Wd Sdng    262
Plywood    185
CmentBd     90
Wd Shng     63
BrkFace     34
Stucco      30
AsbShng     28
Brk Cmn     17
ImStucc     11
Stone        6
AsphShn      3
CBlock       2
Name: Exterior 2nd, dtype: int64

In [39]:
train['Mas Vnr Type'].value_counts()
#Missing about 25 values

None       1218
BrkFace     630
Stone       168
BrkCmn       13
Name: Mas Vnr Type, dtype: int64

In [41]:
train['Mas Vnr Area'].describe()
# Do the areas = 0 correspond with apartments with no outer walls/ none???
# Missing values

count    2029.000000
mean       99.695909
std       174.963129
min         0.000000
25%         0.000000
50%         0.000000
75%       161.000000
max      1600.000000
Name: Mas Vnr Area, dtype: float64

In [45]:
train['Exter Qual'].value_counts()

TA    1247
Gd     697
Ex      81
Fa      26
Name: Exter Qual, dtype: int64

In [46]:
train['Exter Cond'].value_counts()

TA    1778
Gd     215
Fa      49
Ex       7
Po       2
Name: Exter Cond, dtype: int64

In [47]:
train['Foundation'].value_counts()

PConc     926
CBlock    863
BrkTil    221
Slab       34
Stone       5
Wood        2
Name: Foundation, dtype: int64

In [49]:
train['Bsmt Qual'].value_counts()
#Some Missing Values
#Look at description- there should be NA- missing are probably the NAs

TA    887
Gd    864
Ex    184
Fa     60
Po      1
Name: Bsmt Qual, dtype: int64

In [50]:
train['Bsmt Cond'].value_counts()
#Same number of missing values as the basement qual variable- also has no NA

TA    1834
Gd      89
Fa      65
Po       5
Ex       3
Name: Bsmt Cond, dtype: int64

In [51]:
train['Bsmt Exposure'].value_counts()
#3 more missing than other basement variables
# alos supposed to have NA values

No    1339
Av     288
Gd     203
Mn     163
Name: Bsmt Exposure, dtype: int64

In [53]:
train['BsmtFin Type 1'].value_counts()
#Missing values and supposed to have NA values

GLQ    615
Unf    603
ALQ    293
BLQ    200
Rec    183
LwQ    102
Name: BsmtFin Type 1, dtype: int64

In [58]:
train['BsmtFin SF 1'].describe()
#Only missing 1 value, can use this variable to confirm the NA for other basement variable

count    2050.000000
mean      442.300488
std       461.204124
min         0.000000
25%         0.000000
50%       368.000000
75%       733.750000
max      5644.000000
Name: BsmtFin SF 1, dtype: float64

In [56]:
train['BsmtFin Type 2'].value_counts()
# Some missing values, missing the NA

Unf    1749
Rec      80
LwQ      60
BLQ      48
ALQ      35
GLQ      23
Name: BsmtFin Type 2, dtype: int64

In [59]:
train['BsmtFin SF 2'].describe()
#Values are reasonable, can use this variable to confirm NA basement for other variables
# only 1 missing variable

count    2050.000000
mean       47.959024
std       165.000901
min         0.000000
25%         0.000000
50%         0.000000
75%         0.000000
max      1474.000000
Name: BsmtFin SF 2, dtype: float64

In [61]:
train['Bsmt Unf SF'].describe()
# only 1 missing variable

count    2050.000000
mean      567.728293
std       444.954786
min         0.000000
25%       220.000000
50%       474.500000
75%       811.000000
max      2336.000000
Name: Bsmt Unf SF, dtype: float64

In [63]:
train['Total Bsmt SF'].describe()
#Missing 1 value

count    2050.000000
mean     1057.987805
std       449.410704
min         0.000000
25%       793.000000
50%       994.500000
75%      1318.750000
max      6110.000000
Name: Total Bsmt SF, dtype: float64

In [64]:
train['Heating'].value_counts()
#Most values as just GasA

GasA    2018
GasW      20
Wall       6
Grav       5
OthW       2
Name: Heating, dtype: int64

In [65]:
train['Heating QC'].value_counts()
#No missing values

Ex    1065
TA     597
Gd     319
Fa      67
Po       3
Name: Heating QC, dtype: int64

In [66]:
train['Central Air'].value_counts()
#No missing values

Y    1910
N     141
Name: Central Air, dtype: int64

In [67]:
train['Electrical'].value_counts()
#No missing values
#Little variation

SBrkr    1868
FuseA     140
FuseF      35
FuseP       7
Mix         1
Name: Electrical, dtype: int64

In [69]:
train['1st Flr SF'].describe()
# No missing value
#Values probably make sense- 334 sqft is low

count    2051.000000
mean     1164.488055
std       396.446923
min       334.000000
25%       879.500000
50%      1093.000000
75%      1405.000000
max      5095.000000
Name: 1st Flr SF, dtype: float64

In [70]:
train['2nd Flr SF'].describe()
#No missing values
#values of 0 make sense for houses with no second floow

count    2051.000000
mean      329.329108
std       425.671046
min         0.000000
25%         0.000000
50%         0.000000
75%       692.500000
max      1862.000000
Name: 2nd Flr SF, dtype: float64

In [71]:
train['Low Qual Fin SF'].describe()
# No missing values

count    2051.000000
mean        5.512921
std        51.068870
min         0.000000
25%         0.000000
50%         0.000000
75%         0.000000
max      1064.000000
Name: Low Qual Fin SF, dtype: float64

In [72]:
train['Gr Liv Area'].describe()
# Minimum for this variable matches the 1st floor sf min value
# No missing values

count    2051.000000
mean     1499.330083
std       500.447829
min       334.000000
25%      1129.000000
50%      1444.000000
75%      1728.500000
max      5642.000000
Name: Gr Liv Area, dtype: float64

In [77]:
train['Bsmt Full Bath'].value_counts()
#2 missing values
#Should this be ordinal/ categorical instead of numerical?

0.0    1200
1.0     824
2.0      23
3.0       2
Name: Bsmt Full Bath, dtype: int64

In [76]:
train['Bsmt Half Bath'].value_counts()
#0 is by far the most common response, only 126 nonzeros

0.0    1923
1.0     122
2.0       4
Name: Bsmt Half Bath, dtype: int64

In [79]:
train['Full Bath'].value_counts()
# ordinal/ categorical instead of numerical?
# No missing values

2    1096
1     900
3      45
0       8
4       2
Name: Full Bath, dtype: int64

In [ ]:
train['1st Flr SF'].describe()

Variables that should be dropped
- 'Utilities': only 2 values that are not "AllPub"- no predictive value
- 'Alley': Too many missing variables, no way to predict this variable or handle the missing values 
- 'Condition 2': only about 26 values that are non-Norm
- 'Roof Matl': Only 26 values that are varied
- 'Heating'


Variables that maybe should be dropped:
- Land Slope: Only ~100 out of 2000 observations that are not Gtl (But for the non Gtl, is this signficant?, especially severe slope?)
- Central Air: Only 141 Ns, but would that be significant?
- 'Electrical'
- 'Bsmt Half Bath'

Type errors:
- MS SubClass

Type decisions:
-Ordinal variables such as:
- OverallQual
- OverallCond